contracts/core/Allo.sol

```diff
@@ -1,79 +1,368 @@
+
+    /// ==========================
+    /// === Storage Variables ====
+    /// ==========================
+
+    /// @notice Fee percentage
+    /// @dev 1e18 = 100%, 1e17 = 10%, 1e16 = 1%, 1e15 = 0.1%
+    uint256 public feePercentage;
+
+    /// @notice Incremental index
+    uint256 private _poolIndex;
+
+    /// @notice Nonce for cloning strategies
+    uint256 private _nonce;
+
+    /// @notice Allo treasury
+    address payable public treasury;
+
+    /// @notice Registry of pool creators
+    Registry public registry;
+
+    /// @notice Pool.id -> Pool
+    mapping(uint256 => Pool) public pools;
+
+    /// @notice Strategy -> bool
+    mapping(address => bool) public approvedStrategies;
+
+    /// @notice Strategy -> bool
+    mapping(address => bool) public usedStrategies;
+
+    /// ======================
+    /// ======= Events =======
+    /// ======================
+
+    event PoolCreated(
+        uint256 indexed poolId,
+        bytes32 indexed identityId,
+        address allocationStrategy,
+        address distributionStrategy,
+        address token,
+        uint256 amount,
+        Metadata metadata
+    );
 
-    // Public getter on the pools mapping
-    /// @notice calls out to the registry to get the identity metadata
-    function getPoolInfo(uint256 _poolId) external view returns (PoolData memory, string memory) {
-        // Implement the function here
+    event PoolMetadataUpdated(uint256 indexed poolId, Metadata metadata);
+
+    event PoolFunded(uint256 indexed poolId, uint256 amount, uint256 fee);
+
+    event PoolClosed(uint256 indexed poolId);
+
+    event TreasuryUpdated(address treasury);
+
+    event FeeUpdated(uint256 fee);
+
+    event RegistryUpdated(address registry);
+
+    /// ====================================
+    /// =========== Intializer =============
+    /// ====================================
+
+    /// @notice Initializes the contract after an upgrade
+    /// @dev During upgrade -> an higher version should be passed to reinitializer
+    /// @param _registry The address of the registry
+    /// @param _treasury The address of the treasury
+    /// @param _feePercentage The fee percentage
+    function initialize(address _registry, address payable _treasury, uint256 _feePercentage) public reinitializer(1) {
+        _initializeOwner(msg.sender);
+
+        registry = Registry(_registry);
+        treasury = _treasury;
+        feePercentage = _feePercentage;
+
+        emit RegistryUpdated(_registry);
+        emit TreasuryUpdated(_treasury);
+        emit FeeUpdated(_feePercentage);
+
+        __Multicall_init();
+    }
+
+    /// ====================================
+    /// =========== Modifier ===============
+    /// ====================================
+
+    modifier onlyPoolAdmin(uint256 _poolId) {
+        if (!registry.isOwnerOrMemberOfIdentity(pools[_poolId].identityId, msg.sender)) {
+            revert UNAUTHORIZED();
+        }
```

Comment:
I know this is still an open discussion. I'm leaning that it would be better to store role data for each pool separately rather than assuming same as identity, but don't feel super strongly. It's more of a product question. Let's talk about it on the call.

Reply:
Currently this is only used for changing Metadata, but keep in mind it will be called from strategies to control everything — approving applications, submitting merkle root for claims, etc.

Reply:
As per discussion -> we'd handle pool permission by having roles for pool admin via OZ roles

Reply:
fixed via 7f8ce4ad29ea7aede1a61ea0a7cc3f381b6ad061

Reply:
fixed via 7f8ce4ad29ea7aede1a61ea0a7cc3f381b6ad061